pennylane/templates/subroutines/prepselprep.py

```diff
@@ -0,0 +1,122 @@
+import copy
+import itertools
+import numpy as np
+
+import pennylane as qml
+from pennylane import math as qmlmath
+from pennylane.operation import Operation
+
+class PrepSelPrep(Operation):
+    def __init__(self, lcu, control, id=None):
+        coeffs, ops = lcu.terms()
+        control = qml.wires.Wires(control)
+        self.hyperparameters["lcu"] = lcu
+        self.hyperparameters["coeffs"] = coeffs
+        self.hyperparameters["ops"] = tuple(ops)
+        self.hyperparameters["control"] = control
+
+        if any(
+            control_wire in qml.wires.Wires.all_wires([op.wires for op in ops])
+            for control_wire in control
+        ):
+            raise ValueError("Control wires should be different from operation wires.")
+
+
+        target_wires = qml.wires.Wires.all_wires([op.wires for op in ops])
+        self.hyperparameters["target_wires"] = target_wires
+
+
+        all_wires = target_wires + control
+        super().__init__(*self.data, wires=all_wires, id=id)
+
+    def _flatten(self):
+        return (self.lcu), (self.control)
+
+    @classmethod
+    def _unflatten(cls, data, metadata) -> "PrepSelPrep":
+        print(data)
+        return cls(data, metadata)
+
+    def __repr__(self):
+        return f"PrepSelPrep(coeffs={tuple(self.coeffs)}, ops={self.ops}, control={self.control})"
+
+    def map_wires(self, wire_map: dict) -> "Select":
+        new_ops = [o.map_wires(wire_map) for o in self.hyperparameters["ops"]]
+        new_control = [wire_map.get(wire, wire) for wire in self.hyperparameters["control"]]
+        new_lcu = qml.dot(self.hyperparameters["coeffs"], new_ops)
+        return PrepSelPrep(new_lcu, new_control)
+
+    def decomposition(self):
+        return self.compute_decomposition(self.hyperparameters['coeffs'], self.ops, self.control)
+
+    @staticmethod
+    def compute_decomposition(coeffs, ops, control):
+        interface = qml.math.get_interface(coeffs)
+        normalized_coeffs = (np.sqrt(coeffs) / qml.math.norm(np.sqrt(coeffs)))
```

Comment:
Please be careful with using numpy on a differentiable object, created as `coeffs = pnp.array([1/2, 1/2], requires_grad=True)` . You can use functions imported from the [qml.math ](https://docs.pennylane.ai/en/stable/code/qml_math.html) module instead. This model provides functions for framework-agnostic tensor and array manipulation.

Reply:
Yes, I think that's the source of the bug. The problem is that I don't see an equivalent to `np.sqrt`. There's `qml.sqrt_matrix` but that won't work on a 1d array. I might be able to write one that looks similar to `qml.norm` unless there is a better way to do this.

Reply:
Did you try `coeffs**0.5`?

Reply:
That won't work since coeffs is a list of autograd ArrayBoxes. Maybe I could map **0.5 over the list, but I'm not sure how to do that in a way that is agnostic to the framework.

Reply:
If the problem still exist, you might try in your test a Hamiltonian constructed as:

`H = qml.ops.LinearCombination(coeffs, [qml.Y(0), ...] )`

Reply:
It didn't work on that Hamiltonian either. The code `qml.norm(coeffs)` gives the same error.

`TypeError: loop of ufunc does not support argument 0 of type ArrayBox which has no callable sqrt method`

Here is the problem. After `coeffs = pnp.array([1/2, 1/2], requires_grad=True)` the type of coeffs is `pennylane.numpy.tensor.tensor`. After calling  `qml.grad(circuit)(coeffs, ops)` inside the gradient function the type of coeffs is `autograd.numpy.numpy_boxes.ArrayBox`.

Either I need to find a way to stop the type conversion, or I need a way to call sqrt on an ArrayBox, and then need to update `qml.norm` to use this version of sqrt.

Reply:
The problem could also be with passing both `coeffs, ops` in `qml.grad(circuit)(coeffs, ops)`. Please try this and post the result here:

1. Update compute_decomposition with `normalized_coeffs = (qml.math.sqrt(coeffs) / qml.math.norm(qml.math.sqrt(coeffs)))`

2. Run the gradients with a circuit like:

```
dev = qml.device("default.qubit")

coeffs = pnp.array([1/2, 1/2], requires_grad=True)
ops = [qml.Identity(1), qml.PauliZ(1)]


@qml.qnode(dev)
def circuit(coeffs):
    lcu = qml.ops.LinearCombination(coeffs, ops)
    qml.PrepSelPrep(lcu, control=0)
    return qml.expval(qml.Z(1))

qml.grad(circuit)(coeffs)
```
If the problem is still there, you can leave the gradient tests for now and make the rest of the PR review-ready.

Reply:
That fixed the issue. The gradient tests are running for autograd, JAX. I have a running test for what I think is JAX Jit, but I don't understand that part of the code well enough to say for sure. 

Other than cleanup, there's one issue I need to resolve. In the `assert_valid` test the `_check_pytree` test fails when JAX is available, but passes otherwise.